source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,300 @@ namespace EnvoyDefault {
+  }
+
+  if (!is_valid) {
+    auto details = c == '_' ? UhvResponseCodeDetail::get().InvalidUnderscore
+                            : UhvResponseCodeDetail::get().InvalidCharacters;
+    return {RejectAction::Reject, details};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateRequestHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateGenericHeaderValue(const HeaderString& value) {
+  // Verify that the header value is valid.
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-value    = *( field-content / obs-fold )
+  // field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
+  // field-vchar    = VCHAR / obs-text
+  // obs-text       = %x80-FF
+  //
+  // VCHAR          =  %x21-7E
+  //                   ; visible (printing) characters
+  const auto& value_string_view = value.getStringView();
+  bool is_valid = true;
+
+  for (auto iter = value_string_view.begin(); iter != value_string_view.end() && is_valid; ++iter) {
+    is_valid &= testChar(kGenericHeaderValueCharTable, *iter);
+  }
+
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidCharacters};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateContentLengthHeader(const HeaderString& value) {
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2:
+  //
+  // Content-Length = 1*DIGIT
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  auto buffer_end = value_string_view.data() + value_string_view.size();
+
+  std::uint64_t int_value{};
+  auto result = std::from_chars(value_string_view.begin(), value_string_view.end(), int_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
-::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
-HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap&) {
-  return RequestHeaderMapValidationResult::success();
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateHostHeader(const HeaderString& value) {
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-5.4,
+  // and RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2:
+  //
+  // Host       = uri-host [ ":" port ]
+  // uri-host   = IP-literal / IPv4address / reg-name
```

Comment:
There is issue related to invalid IPv6 address https://github.com/envoyproxy/envoy/issues/22859, seems can be validate here.

Reply:
As of right now this PR only validates the characters of an IPv4/IPv6 address and not the syntax. I could add that here but I'm not sure it's in scope for this initial PR. @yanavlasov thoughts?

Reply:
This basically what @yanavlasov said at https://github.com/envoyproxy/envoy/pull/22537/files/07627c81b5e4d2474faabe8b0315bbb38117f9f0#r957748243 (found that comment after I commented this one)

Seems fine to fix that in this PR or a separate PR for me. just thought this is where we should fix the issue.

Reply:
maybe a TODO here?